### PR TITLE
fix: meeting ended and started webhooks trigger

### LINKED ISSUE #14685: [CAL-3505] Issue with Webhooks (Meeting started/ended triggers)
> Hello!
> 
> Currently, after communicating with Zapier support, we have received information from them indicating that the error is likely on your side:
> 
> "We've checked the bug report filed for the Cal.com team and there is no update yet.
> ​
> As Ian mentioned, I might suggest you check with the Cal.com support team, if there might be some settings in Cal.com or some differences with particular meetings that could be preventing Cal.com from sending these notifications to Zapier. "
> 
> The situation is as follows:
> 
> - When scheduling a meeting, information is pulled from Calcom into our CRM ("Trial booked" status).
> - When canceling a meeting or successfully completing a meeting, information from Calcom is not pulled into our CRM (statuses "Canceled" and "Past").
> 
> That is, with the standard "Meeting Ended" event, information from Calcom is only transferred in 5-10% of cases.
> Since April 14th, only 4 users have had information pulled: 1799211, 1777800, 1763228, 1763963.
> 
> Please verify the information and provide a response as soon as possible.

<sub>[CAL-3505](https://linear.app/calcom/issue/CAL-3505/issue-with-zapier)</sub>

### PR DESCRIPTION
**Pull request by CarinaWolli**
## What does this PR do?

This fixes several issues with the Meeting ended and Meeting started trigger. The original issue was that if one booking was canceled, all scheduled triggers for that webhook, also the ones that were not related to the canceled booking, were deleted. This is why the meeting ended/started trigger didn't always trigger. 

Cleans up some old code. Since we use `WebhookScheduledTriggers` with a corn job, `scheduledJobs` are not needed anymore. 

Fixes #14685

## How should this be tested?

- Make sure you have existing bookings
- Create webhook with meeting ended and meeting started trigger
- See that WebhookScheduledTriggers are created for all bookings for each trigger 
- Cancel one of the bookings 
- See that only the WebhookScheduledTriggers related to the booking were related

You can also test: 
- delete webhook 
- activate/deactivate webhook
- edit webhook triggers

### CODE REVIEW

File: packages/trpc/server/routers/viewer/webhook/edit.handler.ts [hunk 1]
```diff
@@ -40,7 +43,12 @@ export const editHandler = async ({ input, ctx }: EditOptions) => {
     data,
   });
 
-  await updateTriggerForExistingBookings(webhook, webhook.eventTriggers, updatedWebhook.eventTriggers);
+  if (data.active) {
+    const activeTriggersBefore = webhook.active ? webhook.eventTriggers : [];
+    await updateTriggerForExistingBookings(webhook, activeTriggersBefore, updatedWebhook.eventTriggers);
+  } else if (!data.active && webhook.active) {
+    await deleteWebhookScheduledTriggers({ webhookId: webhook.id });
```
Comment (CarinaWolli) [inline packages/trpc/server/routers/viewer/webhook/edit.handler.ts:50]: when turning webhook off all webhookScheduledTriggers need to be deleted

File: packages/trpc/server/routers/viewer/webhook/edit.handler.ts [hunk 2]
```diff
@@ -40,7 +43,12 @@ export const editHandler = async ({ input, ctx }: EditOptions) => {
     data,
   });
 
-  await updateTriggerForExistingBookings(webhook, webhook.eventTriggers, updatedWebhook.eventTriggers);
+  if (data.active) {
+    const activeTriggersBefore = webhook.active ? webhook.eventTriggers : [];
```
Comment (CarinaWolli) [inline packages/trpc/server/routers/viewer/webhook/edit.handler.ts:47]: If webhook was not active before, then there were no active triggers even when the webhook has triggers

File: packages/features/bookings/lib/handleCancelBooking.ts [hunk 3]
```diff
@@ -441,24 +443,27 @@ async function handler(req: CustomRequest) {
...
       } else if (reminder.method === WorkflowMethods.WHATSAPP) {
-        deleteScheduledWhatsappReminder(reminder.id, reminder.referenceId);
+        workflowReminderPromises.push(deleteScheduledWhatsappReminder(reminder.id, reminder.referenceId));
       }
-    });
+    }
+  }
+
+  Promise.all([...webhookTriggerPromises, ...workflowReminderPromises]).catch((error) => {
+    log.error("An error occurred when deleting workflow reminders and webhook triggers", error);
```
Comment (Udit-takkar) [inline packages/features/bookings/lib/handleCancelBooking.ts:466]: no await here ?

File: apps/web/test/utils/bookingScenario/expects.ts [hunk 4]
```diff
@@ -288,20 +288,22 @@ export function expectWebhookToHaveBeenCalledWith(
...
-    parsedBody.payload.metadata.videoCallUrl = parsedBody.payload.metadata.videoCallUrl
-      ? parsedBody.payload.metadata.videoCallUrl
-      : parsedBody.payload.metadata.videoCallUrl;
-  }
+  if (parsedBody.payload) {
+    if (parsedBody.payload.metadata?.videoCallUrl) {
+      parsedBody.payload.metadata.videoCallUrl = parsedBody.payload.metadata.videoCallUrl
+        ? parsedBody.payload.metadata.videoCallUrl
+        : parsedBody.payload.metadata.videoCallUrl;
+    }
```
[the remark this answers is not in the document]
Comment (CarinaWolli) [reply apps/web/test/utils/bookingScenario/expects.ts:296]: good question, this code was already there like that 😅 I guess we can remove that

File: apps/web/playwright/webhook.e2e.ts [hunk 5]
```diff
@@ -543,6 +544,131 @@ test.describe("BOOKING_RESCHEDULED", async () => {
...
+        webhook: {
+          select: {
+            userId: true,
+          },
+        },
+        startAfter: true,
+      },
+    });
+
+    expect(scheduledTriggersAfterDisabling.length).toBe(0);
```
Comment (zomars) [inline apps/web/playwright/webhook.e2e.ts:668]: We're receiving a 2 in here:

image

**Review by zomars — APPROVED** [review]
Nice job @CarinaWolli. Seems like the failed e2e is an old one. LGTM.

File: packages/features/webhooks/lib/handleWebhookScheduledTriggers.ts [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        !job.payload || jsonParse(job.payload) ? "application/json" : "application/x-www-form-urlencoded",
+    };
+
+    if (webhook) {
+      headers["X-Cal-Signature-256"] = createWebhookSignature({ secret: webhook.secret, body: job.payload });
+    }
+    fetchPromises.push(
+      fetch(job.subscriberUrl, {
+        method: "POST",
+        body: job.payload,
```
Comment (hussamkhatib) [inline packages/features/webhooks/lib/handleWebhookScheduledTriggers.ts:65]: Isn't the payload of this form ?  
Check https://github.com/calcom/cal.com/issues/15033
```
    "triggerEvent": "",
    "createdAt": "",
    "payload":  {} 
}
```

**Pull request merged**